### ISSUE TITLE
Webservice sensor age differs from xDrip overview

### ISSUE THREAD

**Issue by Paulusma** [NightscoutFoundation/xDrip#1135]
xDrip REST Webservice SVG has an option sensor=Y that will include sensor age in the result.

For a (Libre 1) sensor inserted about 3 days ago:
- overview shows: 3.1d
- at the same time, webservice shows 1d 0h

I did some digging and found that the overview gets its info from preference "nfc_sensor_age" while the webservice receives its value from Sensor.started_at.

What is the difference between both values, and which one should be used to determine end of sensor lifetime?

**Comment by Navid200** [comment]
@Paulusma Would you please close this issue if your question has already been answered?
If not, you may get answers in the xDrip support group on facebook.
https://www.facebook.com/groups/xDripG5
Thanks

**Comment by Paulusma** [comment]
since nobody seems to care

**Comment by mk886714** [comment]
Da ist sicherlich "NFC" zu bevorzugen, jedoch habe ich festgestellt: wenn es um libre geht kommt sehr träge/selten eine Reaktion.

**Comment by Navid200** [comment]
There are too many open issues.  It's a demoralizing situation, for a developer, to work in.  It implies that they are doing a sub-optimal job.  The fact is that the developers have done a superb job.

A user should first consult with other users who may be able to address the concern.
A use, currently, ahs two options for this.  One is facebook:  https://www.facebook.com/groups/xDripG5
The other is:https://gitter.im/jamorham/xDrip-plus

If the user concludes that the other users have a similar experience, then, the user needs to decide:  Can I use xDrip as is and get what I need even if it may not be an elegant solution?  If the answer is yes, the user should not open an issue.
Only if the answer is "no, I cannot get what I need from xDrip as is", should the user open an issue.

**Comment by tolot27** [comment]
@Paulusma Please can you test whether this bug still persist? `Sensors.started_at` should contain the correct value because it is read from the database.

**Comment by Navid200** [comment]
Closing due to inactivity

**Comment by tolot27** [comment]
@jwoglom Did you encounter this problem, too?

**Comment by jwoglom** [comment]
I hadn't noticed this specific behavior before, but there is (at least on the G5/G6 side) a distinguishment between known age of the sensor and the internal xDrip db state of the sensor, and if a similar bug existed on the Libre side that would explain this behavior. 

On further investigation, it appears that the web service is calling [SensorStatus.status()](https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SensorStatus.java#L16) to get this information which is identical to the "Sensor Start" timestamp on the Classic status page. This data is obtained from the latest time you hit "Start Sensor" inside xDrip, which stores a treatment in the xDrip db. In my case using a G6 sensor that I initiate from a tandem pump instead of xDrip, this timestamp is super out of date (in the hundreds of days) as a result. (I have a PR that I'll upload soon-ish which causes xDrip to store a sensor start date for G5/G6 sensors whenever it detects a new sensor being initialized from another source, which should fix this specific problem.)

For G5/G6 that is separate from the "Sensor Status" field on the G5/G6 status page which reads from transmitter data, aka it calculates how many days/hours ago the sensor was started on that transmitter.

The nfc_sensor_age pref field only appears to be read for Libre sensors. I'm not familiar with that part of the code but I presume it's a similar distinction, where it's something obtained from the Libre end by communicating with the Libre, rather than checking the last created "sensor start" treatment in the xDrip db.

That would suggest that the originally reported bug here -- where overview (the specific place unspecified, but let's presume on the Libre status page not the general status page) shows some internal Libre state, while the webservice shows the xDrip database state of the last start it's aware of (which is what you'd also see on the classic status page).

**Comment by Navid200** [comment]
I confirm that if I disable collection on my main phone and start collection on a phone with xDrip just installed on, after the connection is established between the transmitter with the existing ongoing session on it and the new phone, there is a difference between the age of the sensor on the two system status pages.
The one on the G5/G6 status page always comes from the transmitter and is always correct.  The one on the classic status page, in case of the example, shows a few minutes because it goes by what xDrip knows about the sensor not what the transmitter knows.

We can change xDrip so that the classic status page would also get its info from the transmitter.  But, I don't think this is such a big deal.
Is it?

A better solution may be to rename what classic status page shows to "xDrip Sensor Record" instead of the current "Sensor Start".
Can you think of a better wording?
Then, one of us can open a PR to make that request.
What do you think?

**Comment by Navid200** [comment]
What is this bug?
How can it be recreated?  What are the steps?

**Comment by tolot27** [comment]
@jwoglom Please can you link the PR, if any?

I prefer to see the true sensor age at the status page.

@tzachi-dar Do you have any idea how to fix this issue for Libre sensors?

**Comment by Navid200** [comment]
> I prefer to see the true sensor age at the status page.

You do see the sensor age on the G5/G6 status page.  For example, 3.9 days.

What you see on the classic status page is when xDrip started the sensor, which is the time and date.

The two pages don't show the same thing.  They are not supposed to.  But, if xDrip starts the sensor, the two values correspond to each other although not the same.  One is a length of time while the other is a time and date.

If xDrip didn't start the sensor, the G5/G6 status page shows exactly the same thing as before because the source of the information is the transmitter.
In that case, the classic status page shows when xDrip established connectivity to the session in progress.  

This is very useful for debugging someone else's system.  By just seeing those two pages, I can find out a lot about what's going on without having to ask 20 different questions from the user.

It would be nice to know why there is anything wrong with any of this before we start planning to change any of it.

**Comment by Navid200** [comment]
Any objection in converting this to a discussion?

I just tested and was not able to create a discrepancy between the G5/G6 status page and the classic status page.
I believe one of the recent (12 months) updates may have changed the behavior.

**Comment by tolot27** [comment]
A PR should be simple to address the initial problem that the WebService uses the wrong values for Libre and Dexcom sensor age.  @jwoglom Please can you create the [announced](https://github.com/NightscoutFoundation/xDrip/issues/1135#issuecomment-844736163) PR?

> A better solution may be to rename what classic status page shows to "xDrip Sensor Record" instead of the current "Sensor Start".
> Can you think of a better wording?

Changing this is out of the scope of this issue. But we can create a separate PR for clarification. I suggest "Sensor connected" or "Sensor detected" which means the first time the xDrip instance in use connected to a (running) sensor and my have started it if it was a new sensor. 

IMHO, there is no need to turn it into a discussion because just we developers think about a reasonable solution. Hence, this is the right place to discuss this.

**Comment by Navid200** [comment]
Of course, like always, it's a good idea to test with the latest xDrip before doing anything.
I was surprised to find out xDrip behaved very differently than it had before for me.
The originally reported problem may not exist any longer.

**Comment by tolot27** [comment]
> The originally reported problem may not exist any longer.

In that case it still exists because nobody touched

https://github.com/NightscoutFoundation/xDrip/blob/master/app/src/main/java/com/eveningoutpost/dexdrip/UtilityModels/SensorStatus.java#L20

**Comment by Navid200** [comment]
There was a time xDrip acted as a CGM dealing with raw values and performing calibrations.  At that time, the xDrip session had a significance and when it started mattered a lot.

With G6, we used native algorithm meaning we allowed the transmitter to do all the  work and xDrip just reported the values reported by the transmitter with no modifications.
With G7, this trend has continued.  There is absolutely no significance in when an xDrip internal session starts.

I don't know anything about Libre though.
But. my understanding is that we do not run Libre 3 locally.

I am going to close this.  If this conversation needs to continue, it can even if the issue is closed.  And we can reopen if needed.
But, on the Dexcom side, there is absolutely no reason to worry about the xDrip session start time.
